---
title: "class1-c"
author: "Sujatha Jagannathan"
date: "February 6, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
<br/>  

## Data and variable types
Before we start playing around with data, it is important that you are familiar with the following basics.

<br/>  

#### Computations in R
R can be used as an ordinary calculator. Here are some examples.  

Remember **#** is the comment character. The comments give details about the operations in case they are not clear.
```r{}
2 + 3 * 5       # Note the order of operations.
log(10)        # Natural logarithm with base e
5^2            # 5 raised to the second power
3/2            # Division
sqrt(16)      # Square root
abs(3-7)      # Absolute value of 3-7
pi             # The number
exp(2)        # exponential function
# This is a comment line
```

<br/>  

$\color{red}{\text{Your turn...}}$ Try some calculations.  

<br/>  

#### Data structures
R has multiple data structures. If you are familiar with excel you can think of data structures as building blocks of a table and the table itself, and a table is similar to a sheet in excel. Most of the time you will deal with tabular data sets, you will manipulate them, take sub-sections of them. It is essential to know what are the common data structures in R and how they can be used. R deals with named data structures, this means you can give names to data structures and manipulate or operate on them using those names.

<br/>  

![data structures](https://devopedia.org/images/article/46/9213.1526125966.png)
Source: Devopedia

<br/>  

##### Vectors
Vectors are one the core R data structures. It is basically a list of elements of the same type (numeric,character or logical). Later you will see that every column of a table will be represented as a vector. R handles vectors easily and intuitively. You can create vectors with c() function, however that is not the only way. The operations on vectors will propagate to all the elements of the vectors.

```r{}
x <- c(1,3,2,10,5)    #create a vector named x with 5 components
x = c(1,3,2,10,5)  # alternatively you can use '=' though I don't recommend it.
x
y <- 1:5              #create a vector of consecutive integers y
y + 2                 #scalar addition
2*y                 #scalar multiplication
y^2                 #raise each component to the second power
2^y                 #raise 2 to the first through fifth power
y                   #y itself has not been unchanged
y <- y*2
y                   #it is now changed
r1 <- rep(1,3)        # create a vector of 1s, length 3
length(r1)           #length of the vector
class(r1)            # class of the vector
r1[1]            # first element of r1
r1[2]            # second element of r1

r2 <- c("x", "y", "z") #this creates a character vector of length 3
```

<br/>  

$\color{red}{\text{Your turn...}}$ Create a new vector `r3` containing the names of yourself, and the two people on either side of you. 

<br/>  

##### Matrices
A __matrix__ refers to a numeric array of rows and columns. You can think of it as a stacked version of vectors where each row or column is a vector. One of the easiest ways to create a matrix is to combine vectors of equal length using `cbind()`, meaning 'column bind'.

```r{}
x <- c(1,2,3,4)      # c is for combine
y <- c(4,5,6,7)
m1 <- cbind(x,y)
m1
t(m1)                # transpose of m1
dim(m1)              # 2 by 5 matrix
```


You can also creat a matrix by listing the elements using `matrix()`:
```r{}
m2 <- matrix(c(1,3,2,5,-1,2,2,3,9),nrow=3)
m2
```
<br/>  

__Matirices__, __data frames__, and __tibbles__ (the latter two will be covered next) are all tabular data structures. You can subset them using **[ ]** and providing desired rows and columns to subset. Here is how that works conceptually:
![alt text](http://compgenomr.github.io/book/Intro2R_data/nonR_figures/slicingDataFrames.png "slicing/subsetting")
<br/>  

*slicing/subsetting of a matrix and data frame*.There are additional ways to subset data frames, see the next section.

<br/>  


##### Data Frames
A data frame is more general than a matrix, in that different columns can have different modes (numeric, character, factor, etc.). A data frame can be constructed by data.frame() function. For example, we illustrate how to construct a data frame from genomic intervals or coordinates.

```r{}
chr <- c("chr1", "chr1", "chr2", "chr2")
strand <- c("-", "-", "+", "+")
start <- c(200,4000,100,400)
end <-c(250,410,200,450)
mydata <- data.frame(chr, start, end, strand)
names(mydata)
#change column names
names(mydata) <- c("A", "B", "C", "D")
mydata 
# here you can assign names  (in caps in this example) to each list that will be combined to make a dataframe. These list names will be transferred to the column name.
mydata <- data.frame(CHROM=chr,
                     START=start,
                     STOP=end,
                     STRAND=strand)
mydata
```

There are a variety of ways to extract the elements of a data frame. You can extract certain columns using column numbers or names, or you can extract certain rows by using row numbers. You can also extract data using logical arguments, such as extracting all rows that has a value in a column larger than your threshold.

```r{}
mydata[,2:4] # columns 2,3,4 of data frame
mydata[,c("CHROM","START")] # columns chr and start from data frame
mydata$START # variable start in the data frame
mydata[c(1,3),] # get 1st and 3rd rows
mydata[mydata$START>400,] # get all rows where start>400
```

<br/>  

##### Tibbles
A __tibble__, or `tbl_df`, is a modern reimagining of the data.frame, keeping what time has proven to be effective, and throwing out what is not. Tibbles are data.frames that are lazy and surly: they do less (i.e. they don't change variable names or types, and don't do partial matching) and complain more (e.g. when a variable does not exist). This forces you to confront problems earlier, typically leading to cleaner, more expressive code. Tibbles also have an enhanced `print()` method which makes them easier to use with large datasets containing complex objects.

If you are new to tibbles, the best place to start is the [tibbles chapter](http://r4ds.had.co.nz/tibbles.html) in *R for data science*.

Create a tibble from an existing object with `as_tibble()`:

```r{}
library(tibble)
as_tibble(iris)
```

This will work for reasonable inputs that are already data.frames, lists, matrices, or tables.

You can also create a new tibble from column vectors with `tibble()`:

```r{}
tibble(x = 1:5, y = 1, z = x ^ 2 + y)
```

`tibble()` does much less than `data.frame()`: it never changes the type of the inputs (e.g. it never converts strings to factors!), it never changes the names of variables, it only recycles inputs of length 1, and it never creates `row.names()`. You can read more about these features in the vignette, `vignette("tibble")`.

<br/>  

##### Lists
An ordered collection of objects (components). A list allows you to gather a variety of (possibly unrelated) objects under one name.
```r{}
# example of a list with 4 components
# a string, a numeric vector, a matrix, and a scalar
w <- list(name="Fred",
       mynumbers=c(1,2,3),
       mymatrix=matrix(1:4,ncol=2),
       age=5.3)
w
```
You can extract elements of a list using the **[[]]** convention using either its position in the list or its name.
```r{}
w[[3]] # 3rd component of the list
w[["mynumbers"]] # component named mynumbers in list
w$age
```

<br/>  

##### Factors
Factors are used to store categorical data. They are important for statistical modeling since categorical variables are treated differently in statistical models than continuous variables. This ensures categorical data treated accordingly in statistical models.
```r{}
features=c("promoter","exon","intron")
f.feat=factor(features)
```
Important thing to note is that when you are reading a data.frame with read.table() or creating a data frame with **data.frame()** character columns are stored as factors by default, to change this behavior you need to set **stringsAsFactors=FALSE** in **read.table()** and/or **data.frame()** function arguments.

<br/>  

### Data types
There are four common data types in R, they are **numeric**, **logical**, **character** and **integer**. All these data types can be used to create vectors natively.
```r{}
#create a numeric vector x with 5 components
x <- c(1,3,2,10,5)
x

#create a logical vector x
x <- c(TRUE, FALSE, TRUE)
x

# create a character vector
x <- c("sds","sd","as")
x

# create an integer vector
x <- c(1L,2L,3L)
x
```

<br/>  

### Querying data

R provides many functions to examine features of vectors and other objects, for example

`class()` - what kind of object is it (high-level)?
`str()` - what is the structure of the object?
`typeof()` - what is the objectâ€™s data type (low-level)?
`length()` - how long is it? What about two dimensional objects?
`attributes()` - does it have any metadata?

```r{}
# determine which of the data types 'x' actually is using the class() function
class(x)

# typeof() is another function to find data type
typeof(x)

# use is.? to confirm data type
is.integer(x)
is.tibble(iris)

# one data type can be coerced to another using as.<class_name>()
as.character(x)
```

<br/>  

$\color{red}{\text{Your turn...}}$ Use the above functions to query `mydata` that you created earlier. 

<br/>  

## Acknowledgements

The material for this class was heavily borrowed from: http://compgenomr.github.io/book/introduction-to-r.html