---
title: "Practical Computational Biology: R"
author: "Sujatha Jagannathan"
date: "February 25, 2019"
output: ioslides_presentation
incremental: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Learning objectives for this course
> The primary learning objectives are to work towards being able to:  

* Retrieve and import genomic data.  
* Prepare data for analysis (“tidy”-ing).  
* Conduct basic exploratory analysis - getting familiar with your dataset.  
* Test specific hypotheses using computational methods.  
* Visualize and communicate your results.  
* Practice reproducible analysis.  

## Today's class outline

- Coding best practices
- Review data types
- In-class exercise on RNAseq dataset
    + Import
    + Tidy (with tidyr)
    + Transform (with dyplyr)
    + Export
- Homework instructions

## Class rules {.build .bigger}

- Ask for help (use the stickie notes)  

- Offer help!   
      + We have a wide range of R experience in this class room  
      + Pair with someone with a different R expertise and help each other  

- Use Google *liberally*  (& the cheatsheets!)

## Coding best practices

>   "Good coding style is like correct punctuation: you can manage without it, butitsuremakesthingseasiertoread."
>   
>   --- Hadley Wickham

## Data types and structures {.smaller}
- Basic data types: numeric, character, date, and logical. 
- Basic data structures: vector, matrix, array, data frame (or, tibble), and list. 
```{r}
knitr::include_graphics("/Users/sjaganna/Desktop/CU-onedrive/OneDrive - The University of Colorado Denver/08-teaching/molb7910/1-class1/images/data-structures.png", dpi = 150)
```
- Data can be queried to understand it better: `class()`, `str()`, etc. 

## Import, tidy, transform, and export an RNA-seq count dataset {.smaller}

- Three data sets
    + RNAseq count data (Dyle *et al*; unpublished; please keep confidential)
    + Experimental metadata
    + Gene annotation
    
<br/>

- Our goal is to import these data, clean up by eliminating unnecessary columns and renaming variables, merge with the gene annotation file to map ensembl ids, summarize transcript-level count data to gene-level (i.e. transform), and then export the file.

## Homework

- Instructions online

- Will involve the use of the gene-level data we exported today 