---
title: "class4-DGE.Rmd"
author: "Sujatha Jagannathan"
date: "February 22, 2019"
output: html_document
---

<br/>  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Preparing R
Load packages  
```{r, include = TRUE, warning=FALSE, message=FALSE}
library("tidyverse")
library("here")
library("edgeR")
```
This should install readr, tidyr, dplyr, purr, stringr, ggplot2, tibble, and forcats.

Set working directory
```{r}
here()
setwd(here())
```

<br/>  

## Differential gene expression using edgeR    

<br/>  

#### Introduction

There are a number of steps in conducting differential gene expression analysis using RNA-seq data. Here, we assume that you have already mapped and counted the reads from an RNA-seq experiment (i. e. fastq data has been mapped to a reference genome and reads that map to a certain gene have been counted using a program such as STAR). Once you have the count data, we will generate matrix containing sample names as column ids and gene names as row ids. This matrix can then be input into edgeR to generate log2 fold change values for each gene along with a p-value and fdr. The code we will use to accomplish this is simple, but does a LOT to your data behind the scenes. 

```{r, echo = FALSE}
knitr::include_graphics(here("4-class4", "images", "rnaseq-pipeline.png"), dpi = 200)
```

<br/>  

#### edgeR pipeline

1. **Input data** as a count matrix (rownames = gene ids, colnames = sample ids)
2. **Define grouping** of replicates. (if you have multiple replicates, you need to specify how they are organized with respect to the different samples)
3. Combine #1 and #2 to **generate a DGElist** using `DGElist(counts=<data>, group=<grouping>)`
4. **Filter** genes with low counts using `data[keep]`
5. **Normalize** the counts post-filtering via `calcNormFactors()`
6. Convert #2 into an **experimental design** using `model.matrix()` for linear modeling. 
7. **Estimate dispersion** using `estimateDisp(<count-matrix>, <design>)`
8. **Fit a generalized linear model** using `glmFit(<output-from-above>, <design>)`
9. **Log-likelihood ratio test** using `glmLRT(<output-from-above>, coef=2)`
10. Find **top differentially expressed genes** using `topTags(<output-from-above$table>, n=10733)`

<br/>  

Let's get started! 

<br/>  

#### Importing data
```{r}
data_gene_level <- read_csv(here("4-class4", 
                                      "data", 
                                     "input",
                                     "data_gene-level.csv"))

```

<br/>  

#### Prepare data
```{r}
geneExpr <- data_gene_level %>% #pipe data to next step
                  select(-X1, -hgnc_symbol) %>% #Remove row# column and gene symbol
                  rename(id = ensembl_gene_id) %>% #rename the ensembl_gene_id to id
                  filter(!is.na(id)) # remove rows with no gene ids. 

geneExpr <- data.frame(geneExpr) #we are converting the geneExpr tibble to a data.frame for a specific reason. Tibbles do not allow a variable to be contained in a row name. Therefore, they don't allow row names at all. But for edgeR to work, we need the count data to be in a matrix with gene names as row names and sample names as column names. 
row.names(geneExpr) <- geneExpr$id # This is where we move the gene id information into row name
x <- geneExpr[, c(2:13)] # then, we get rid of the column containing gene id. Thus, generating out matrix of count data. 
```

<br/>  

#### edgeR analysis
```{r}
# 1. Subsetting the data to only consider 0 vs 14 hour for our fold-change calculations
x <- x[, c(1:3, 10:12)] #keeping 0h and 14h samples only

#Note: You can create as many sets as you want for the different comparisions you want to make, such as x1, x2, and so on. Just make sure to do the following steps also for each of the sets you made.

# 2. Define sample grouping
group <- factor(c(1, 1, 1, 2, 2, 2))

#Note: If you made more than one sets, you need to define their grouping individually as group1, group2, etc. 

# 3. Generate a DGE list
d <- DGEList(counts=x, group=group)

# 4. Filter genes with low counts (require >1 cpm in at least 1/2 of the samples) & normalize
keep <- rowSums(cpm(d)>1) >= 3 # identify the rows to keep that have cpm() > 1
d <- d[keep, , keep.lib.sizes=FALSE] # remove the rows that don't have sufficient cpm

# 5. Calculate normalization factors to scale the raw library sizes.
d <- calcNormFactors(d)

# 6. create a model matrix to allow linear modeling
design <- model.matrix(~group)

# 7. estimate dispersion
d <- estimateDisp(d, design)

# 8. Fit a generalized linear model 
d1 <- glmFit(d, design)

# 9. Conduct Tagwise likelihood ratio tests for DGEGLM
d2 <- glmLRT(d1, coef=2)

# 10. Extract the most differentially expressed genes (or sequence tags) from a test object, ranked either by p-value or by absolute log-fold-change.
DUX4_14h <- topTags(d2, n=10733) 
```

<br/>  

#### Data clean up and export
```{r}
## Assemble final data structure for ploting figures
foldchange_hour14 <- cbind(id = row.names(DUX4_14h$table), DUX4_14h$table)

foldchange <- left_join(data_gene_level, foldchange_hour14, by = c("ensembl_gene_id" = "id"))

foldchange <- unique(foldchange)
```

<br/>  

#### SessionInfo

```{r}
sessionInfo()
```

#### References
Youtube video on edgeR: https://www.youtube.com/watch?v=5tGCBW3_0IA  
edgeR tutorial 1: https://web.stanford.edu/class/bios221/labs/rnaseq/lab_4_rnaseq.html
edgeR tutorial 2: https://gist.github.com/jdblischak/11384914