---
title: "class3-DGE.Rmd"
author: "Sujatha Jagannathan"
date: "February 22, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Some tutorials
https://www.youtube.com/watch?v=5tGCBW3_0IA
https://web.stanford.edu/class/bios221/labs/rnaseq/lab_4_rnaseq.html

## 1. Preparing R
Load packages  
```{r, include = TRUE, warning=FALSE, message=FALSE}
library("tidyverse")
library("skimr")
library("here")
library("edgeR")
library("dplyr")
```
This should install readr, tidyr, dplyr, purr, stringr, ggplot2, tibble, and forcats.

Set working directory
```{r}
here()
setwd(here())
```

## 2. Importing data
```{r}

data_gene_level <- read_csv(here("3-class3", 
                                      "data", 
                                     "input",
                                     "data_gene-level.csv"))
```

## 3. Prepare data
```{r}
geneExpr <- data_gene_level %>% #pipe data to next step
                  select(-X1, -hgnc_symbol) %>% #Remove row# column and gene symbol
                  rename(id = ensembl_gene_id) %>% #rename the ensembl_gene_id to id
                  filter(!is.na(id)) # remove rows with no gene ids. 

geneExpr <- data.frame(geneExpr) #we are converting the geneExpr tibble to a data.frame for a specific reason. Tibbles do not allow a variable to be contained in a row name. Therefore, they don't allow row names at all. But for edgeR to work, we need the count data to be in a matrix with gene names as row names and sample names as column names. 
row.names(geneExpr) <- geneExpr$id # This is where we move the gene id information into row name
x <- geneExpr[, c(2:13)] # then, we get rid of the column containing gene id. Thus, generating out matrix of count data. 
```

## 4. Subset the data to the samples to be compared. 
```{r}
# Subsetting the data to only consider 0 vs 14 hour for our fold-change calculations
x_set <- x[, c(1:3, 10:12)] #14h

#Note: You can create as many sets as you want for the different comparisions you want to make, such as x_set1, x_set2, and so on. Just make sure to do the following steps also for each of the sets you made.
```

## 5. Define sample grouping
```{r}
group_set <- c(1, 1, 1, 2, 2, 2)

#Note: If you made more than one sets, you need to define their grouping individually as group_set1, group_set2, etc. 
```

## 6. Create the design
```{r}
design_set <- model.matrix(~group_set)

#Note: same as above. each set needs its own design. 
```

## 6. Generate a DGE list
```{r}
y_set <- DGEList(counts=x_set, group=group_set)
```

## 7. Filter genes with low counts (require >1 cpm in at least 1/2 of the samples)
```{r}
keep <- rowSums(cpm(y_set)>1) >= 3
y_set <- y_set[keep, , keep.lib.sizes=FALSE]
```

## 8. Normalize
```{r}
y1_set <- calcNormFactors(y_set)
```

## 9. Find differentially expressed genes
```{r}
#estimate dispersion
y2_set <- estimateDisp(y1_set, design_set)

#find DE genes
fit_set1 <- glmFit(y2_set1, design_set1)
lrt.2vs1_set1 <- glmLRT(fit_set1, coef=2) #coef specifies ?
DUX4_4h <- topTags(lrt.2vs1_set1, n=10733) #n specifies number of genes to view
```

## 10. Data clean up and export
```{r}
## Clean up the data table
names(DUX4_4h$table) <- c("hour04_logFC", "hour04_logCPM", "hour04_LR", "hour04_pval", "hour04_fdr")
names(DUX4_8h$table) <- c("hour08_logFC", "hour08_logCPM", "hour08_LR", "hour08_pval", "hour08_fdr")
names(DUX4_14h$table) <- c("hour14_logFC", "hour14_logCPM", "hour14_LR", "hour14_pval", "hour14_fdr")

## Assemble final data structure for ploting figures
hour04 <- cbind(id = row.names(DUX4_4h$table), DUX4_4h$table)
hour08 <- cbind(id = row.names(DUX4_8h$table), DUX4_8h$table)
hour14 <- cbind(id = row.names(DUX4_14h$table), DUX4_14h$table)

library("dplyr")
foldchange <- full_join(hour04, hour08, by = "id")
foldchange <- full_join(foldchange, hour14, by = "id")
ids <- foldchange$id
foldchange <- merge(genes, foldchange, by = "id")
```

```{r}
plot(foldchange$hour04_logCPM, foldchange$hour04_logFC, 
     pch=16, 
     cex=1,
     cex.lab=1.5, 
     cex.axis=1.5, 
     cex.main=1.5, 
     cex.sub=1.5,
     ylim = c(-12, 12),
     col=ifelse(foldchange$hour04_fdr > 0.05, "#10101020",
                ifelse(foldchange$hour04_logFC < -2, "#2C7BB6",
                       ifelse(foldchange$hour04_logFC > 2, "#D71E1D","#AADDA340"))))

plot(foldchange$hour08_logCPM, foldchange$hour08_logFC, 
     pch=16, 
     cex=1,
     cex.lab=1.5, 
     cex.axis=1.5, 
     cex.main=1.5, 
     cex.sub=1.5,
     ylim = c(-12, 12),
     col=ifelse(foldchange$hour08_fdr > 0.05, "#10101020",
                ifelse(foldchange$hour08_logFC < -2, "#2C7BB6",
                       ifelse(foldchange$hour08_logFC > 2, "#D71E1D","#AADDA340"))))

plot(foldchange$hour14_logCPM, foldchange$hour14_logFC, 
     pch=16, 
     cex=1,
     cex.lab=1.5,
     cex.axis=1.5, 
     cex.main=1.5, 
     cex.sub=1.5,
     ylim = c(-12, 12),
     col=ifelse(foldchange$hour14_fdr > 0.05, "#10101020",
                ifelse(foldchange$hour14_logFC < -2, "#2C7BB6",
                       ifelse(foldchange$hour14_logFC > 2, "#D71E1D","#AADDA340"))))

identify(foldchange$hour14_logCPM, foldchange$hour14_logFC,
         labels = foldchange$genes)


smoothScatter(foldchange$hour04_logFC, foldchange$hour14_logFC,
     cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5, 
              colramp=colorRampPalette(c("white", "#e5f5f9", "#99d8c9", "#2ca25f")))
abline(lm(foldchange$hour14_logFC ~ foldchange$hour04_logFC))

smoothScatter(foldchange$hour04_logFC, foldchange$hour08_logFC,
     cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5, 
              colramp=colorRampPalette(c("white", "#e5f5f9", "#99d8c9", "#2ca25f")))
abline(lm(foldchange$hour08_logFC ~ foldchange$hour04_logFC))

smoothScatter(foldchange$hour08_logFC, foldchange$hour14_logFC,
     cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5, 
              colramp=colorRampPalette(c("white", "#e5f5f9", "#99d8c9", "#2ca25f")))
abline(lm(foldchange$hour14_logFC ~ foldchange$hour08_logFC))
```

## Generating UP and DOWN genes
```{r}
up <- nrow(foldchange %>% filter (hour14_logFC > 2 & hour14_fdr < 0.05))
down <- nrow(foldchange %>% filter (hour14_logFC < -2 & hour14_fdr < 0.05))
```