---
title: "ribosome-footprinting-analysis"
author: "Sujatha Jagannathan"
date: "January 8, 2019"
output: html_document
---

## 1. R setup
```{r message=FALSE, warning=FALSE, setup}
suppressMessages(library(tidyverse))
suppressMessages(library(ggpubr))
library("viridis")
library("RColorBrewer")
suppressMessages(library("PerformanceAnalytics"))
library("DT")
library("DataExplorer")
library("GGally")
library("ggthemes")
library("gridExtra")
library("ggfortify")
library("ggpubr")
library("corrplot")
library("matrixStats")
```

## 2. Data input
```{r message=FALSE, warning=FALSE, data-input}
library(readr)
count_summary_rna_010319 <- read_delim("count_summary_rna_010319.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

library(readr)
count_summary_ribo_010319 <- read_delim("count_summary_ribo_010319.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

load("data.rnaseq.silac.static_06072018.Rdata")
mart_export_coding <- read_csv("mart_export coding.txt")
mart_export_nonsense <- read_csv("mart_export nonsense.txt")
biomart <- rbind(mart_export_coding, mart_export_nonsense)
```

## 3. Data clean up
```{r message=FALSE, warning=FALSE, cleanup}
rnaseq <- count_summary_rna_010319
riboseq <- count_summary_ribo_010319

rnaseq <- rnaseq[, c(1, 6:18)]
#fixing the lable mixup

rnaseq <- rnaseq[, c(1:5, 9:11, 12:14, 6:8)]
names(rnaseq) <- c("geneid", "length", "0h_1", "0h_2", "0h_3",  "4h_1", "4h_2", "4h_3", "8h_1", "8h_2", "8h_3",  "14h_1", "14h_2", "14h_3")

riboseq <- riboseq[, c(1, 6:18)]
names(riboseq) <- c("geneid", "length", "0h_1", "0h_2", "0h_3",  "4h_1", "4h_2", "4h_3", "8h_1", "8h_2", "8h_3",  "14h_1", "14h_2", "14h_3")

```

## 4. Data processing

### i) Counting reads before processing
```{r message=FALSE, warning=FALSE, fig.height=6, fig.width=12, counts}
count.rna <- rnaseq %>% select (- geneid, - length) %>% colSums(., na.rm = T)
count.rna.df <- as.data.frame(count.rna)
count.rna.df <- data.frame(sample = rownames(count.rna.df), counts = count.rna)
count.rna.df$sample <- factor(count.rna.df$sample, levels = c("0h_1","0h_2","0h_3","4h_1", "4h_2", "4h_3","8h_1", "8h_2","8h_3", "14h_1","14h_2","14h_3"))

rna.plot <- ggplot(count.rna.df,aes(x = sample, y = counts)) +
  geom_bar(stat="identity") +
  theme_few() +
  labs(ylab("Read Counts")) +
  labs(xlab("Sample")) +
  ggtitle("Read Count plot - RNAseq") +
  coord_flip()

count.ribo <- riboseq %>% select (- geneid, - length) %>% colSums(., na.rm = T)
count.ribo.df <- as.data.frame(count.ribo)
count.ribo.df <- data.frame(sample = rownames(count.ribo.df), counts = count.ribo)
count.ribo.df$sample <- factor(count.ribo.df$sample, levels = c("0h_1","0h_2","0h_3","4h_1", "4h_2", "4h_3","8h_1", "8h_2","8h_3", "14h_1","14h_2","14h_3"))

ribo.plot <- ggplot(count.ribo.df,aes(x = sample, y = counts)) +
  geom_bar(stat="identity") +
  theme_few() +
  labs(ylab("Read Counts")) +
  labs(xlab("Sample")) +
  ggtitle("Read Count plot - RiboSeq") +
  coord_flip()

ggarrange(rna.plot,ribo.plot, labels = c("A", "B"),
          ncol = 2, nrow = 1)
```

### ii) Calculating RPKM
```{r  message=FALSE, warning=FALSE, RPKM}
library(dplyr)
library(tidyr)
rnaseq_rpkm <- rnaseq %>% mutate_at(vars(-geneid, - length), funs((.*1000000/sum(., na.rm = T))/(length*0.001)))
riboseq_rpkm <- riboseq %>% mutate_at(vars(-geneid, - length), funs((.*1000000/sum(., na.rm = T))/(length*0.001)))

rnaseq_rpkm <- rnaseq_rpkm %>% separate(geneid, c("geneid", "extra"), "\\.")
riboseq_rpkm <- riboseq_rpkm %>% separate(geneid, c("geneid", "extra"), "\\.")
rnaseq_rpkm <- rnaseq_rpkm[, c(1, 4:15)]
riboseq_rpkm <- riboseq_rpkm[, c(1, 4:15)]
```


### iii) Counting reads after RPKM measurement

```{r  message=FALSE, warning=FALSE, fig.height=6, fig.width=12, count-rpkm}
rpkm.rna <- rnaseq_rpkm %>% select (- geneid) %>% colSums(., na.rm = T)
rpkm.rna.df <- as.data.frame(rpkm.rna)
rpkm.rna.df <- data.frame(sample = rownames(rpkm.rna.df), counts = rpkm.rna)
rpkm.rna.df$sample <- factor(rpkm.rna.df$sample, levels = c("0h_1","0h_2","0h_3","4h_1", "4h_2", "4h_3","8h_1", "8h_2","8h_3", "14h_1","14h_2","14h_3"))

rna.plot <- ggplot(rpkm.rna.df,aes(x = sample, y = counts)) +
  geom_bar(stat="identity") +
  theme_few() +
  labs(ylab("Read Counts")) +
  labs(xlab("Sample")) +
  ggtitle("Read Count plot - RNAseq RPKM") +
  coord_flip()

rpkm.ribo <- riboseq_rpkm %>% select (- geneid) %>% colSums(., na.rm = T)
rpkm.ribo.df <- as.data.frame(rpkm.ribo)
rpkm.ribo.df <- data.frame(sample = rownames(rpkm.ribo.df), counts = rpkm.ribo)
rpkm.ribo.df$sample <- factor(rpkm.ribo.df$sample, levels = c("0h_1","0h_2","0h_3","4h_1", "4h_2", "4h_3","8h_1", "8h_2","8h_3", "14h_1","14h_2","14h_3"))

ribo.plot <- ggplot(rpkm.ribo.df,aes(x = sample, y = counts)) +
  geom_bar(stat="identity") +
  theme_few() +
  labs(ylab("Read Counts")) +
  labs(xlab("Sample")) +
  ggtitle("Read Count plot - RiboSeq RPKM") +
  coord_flip()

ggarrange(rna.plot,ribo.plot, labels = c("A", "B"),
          ncol = 2, nrow = 1)
```

### iv) Mapping to geneid
```{r  message=FALSE, warning=FALSE, geneid}
names(rnaseq_rpkm) <- c("ensembltranscript","0h_1", "0h_2", "0h_3",  "4h_1", "4h_2", "4h_3", "8h_1", "8h_2", "8h_3",  "14h_1", "14h_2", "14h_3")
names(riboseq_rpkm) <- c("ensembltranscript","0h_1", "0h_2", "0h_3",  "4h_1", "4h_2", "4h_3", "8h_1", "8h_2", "8h_3",  "14h_1", "14h_2", "14h_3")

names(biomart) <- c("ensemblgene", "ensembltranscript", "length", "cdslength", "geneid")

rnaseq_mapped <- right_join(biomart, rnaseq_rpkm, by = "ensembltranscript")
riboseq_mapped <- right_join(biomart, riboseq_rpkm, by = "ensembltranscript")
```
`r datatable(rnaseq_mapped, filter = 'top', options = list(pageLength = 5, autoWidth = TRUE))`.

`r datatable(riboseq_mapped, filter = 'top', options = list(pageLength = 5, autoWidth = TRUE))`.

### v) Calculating fold changes with pseudocount of 0.01
```{r combine-data}
rnaseq_mapped <- rnaseq_mapped %>% 
  mutate(rna_0h = ((`0h_1` +`0h_2` + `0h_3` + 0.01)/3), 
         rna_4h = ((`4h_1` +`4h_2` + `4h_3` + 0.01)/3),
         rna_8h = ((`8h_1` +`8h_2` + `8h_3` + 0.01)/3),
         rna_14h = ((`14h_1` +`14h_2` + `14h_3` + 0.01)/3))

riboseq_mapped <- riboseq_mapped %>% 
  mutate(ribo_0h = ((`0h_1` +`0h_2` + `0h_3` + 0.01)/3),
         ribo_4h = ((`4h_1` +`4h_2` + `4h_3` + 0.01)/3),
         ribo_8h = ((`8h_1` +`8h_2` + `8h_3` + 0.01)/3),
         ribo_14h = ((`14h_1` +`14h_2` + `14h_3` + 0.01)/3))

rnaseq_mapped <- rnaseq_mapped[, c(2, 5, 18:21)]
riboseq_mapped <- riboseq_mapped[, c(2, 5, 18:21)]

rnaseq_mapped <- rnaseq_mapped %>% drop_na(ensembltranscript)
riboseq_mapped <- riboseq_mapped %>% drop_na(ensembltranscript)

rna_ribo_mapped <- full_join(rnaseq_mapped, riboseq_mapped, by = "ensembltranscript")
rna_ribo_mapped <- rna_ribo_mapped %>% 
  mutate(rna.ratio.4h = log2((rna_4h)/(rna_0h)), 
         ribo.ratio.4h = log2((ribo_4h)/(ribo_0h)),
         rna.ratio.8h = log2((rna_8h)/(rna_0h)), 
         ribo.ratio.8h = log2((ribo_8h)/(ribo_0h)), 
         rna.ratio.14h = log2((rna_14h)/(rna_0h)), 
         ribo.ratio.14h = log2((ribo_14h)/(ribo_0h)))

rna_ribo_mapped <- rna_ribo_mapped[, c(2:6, 8:17)]
rna_ribo_mapped <- rna_ribo_mapped %>% drop_na(geneid.x)

data_sub <- rna_ribo_mapped[, c(1, 10:15)]
names(data_sub) <- c("geneid", "rna.ratio.4h", "ribo.ratio.4h", "rna.ratio.8h", "ribo.ratio.8h", "rna.ratio.14h", "ribo.ratio.14h")
data_sub <- data_sub %>% 
  group_by(geneid) %>% 
  summarise(rna.ratio.4h = mean(rna.ratio.4h), 
            ribo.ratio.4h = mean(ribo.ratio.4h), 
            rna.ratio.8h = mean(rna.ratio.8h), 
            ribo.ratio.8h = mean(ribo.ratio.8h),
            rna.ratio.14h = mean(rna.ratio.14h), 
            ribo.ratio.14h = mean(ribo.ratio.14h))
```

### Vi) Merging with silac data
```{r}
data_rna <- data.rnaseq.silac.static %>% select(geneid, iDUX4_logFC)
data_silac <- data.rnaseq.silac.static %>% filter(rep != 3, rep != 4) %>% select(geneid, hl.ratio)
data_rna_ribo_silac <- full_join(data_sub, data_rna, by = "geneid")
data_rna_ribo_silac <- full_join(data_rna_ribo_silac, data_silac, by = "geneid")

data_rna_ribo_silac <- data_rna_ribo_silac %>% unique()
write.csv(data_rna_ribo_silac, file = "data_rna_ribo_silac.csv")
```
`r datatable(data_rna_ribo_silac, filter = 'top', options = list(pageLength = 5, autoWidth = TRUE))`.

## 7. Data exploration
### i) Correlation plots
```{r  message=FALSE, warning=FALSE, fig.height=6, fig.width=12}
## define functions
# Get lower triangle of the correlation matrix
  get_lower_tri<-function(cormat){
    cormat[upper.tri(cormat)] <- NA
    return(cormat)
  }
# Get upper triangle of the correlation matrix
  get_upper_tri <- function(cormat){
    cormat[lower.tri(cormat)]<- NA
    return(cormat)
  }

# correlation between time points at RNA level
  
cormat <- round(cor(rnaseq_rpkm[, c(2:13)], use = "complete.obs"),2)
upper_tri <- get_upper_tri(cormat)

library(reshape2)
melted_cormat <- melt(upper_tri, na.rm = TRUE)
head(melted_cormat)

# Heatmap
library(ggplot2)
rna.plot <- ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name="Pearson\nCorrelation") +
  theme_minimal()+ 
  ggtitle("RNASeq across time points") +
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1)) +
 coord_fixed()

# correlation between time points at ribosome footprint level
cormat <- round(cor(riboseq_rpkm[, c(2:13)], use = "complete.obs"),2)
upper_tri <- get_upper_tri(cormat)

library(reshape2)
melted_cormat <- melt(upper_tri, na.rm = TRUE)
head(melted_cormat)

# Heatmap
library(ggplot2)
ribo.plot <- ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name="Pearson\nCorrelation") +
  theme_minimal()+ 
    ggtitle("RiboSeq across time points") +
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1))+
 coord_fixed()

ggarrange(rna.plot,ribo.plot, labels = c("A", "B"),
          ncol = 2, nrow = 1)
```

### ii) Correlation between all datasets (4 time points x triplicates x RNA|Protein)
```{r  message=FALSE, warning=FALSE, fig.height=6, fig.width=12, exploration4}
# correlation between average counts
cormat <- round(cor(rna_ribo_mapped[, c(2:9)], use = "complete.obs"),2)
upper_tri <- get_upper_tri(cormat)

library(reshape2)
melted_cormat <- melt(upper_tri, na.rm = TRUE)
head(melted_cormat)

# Heatmap
plot1 <- ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name="Pearson\nCorrelation") +
  theme_minimal()+ 
      ggtitle("Average RNA- and Ribo-Seq across time points") +
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1))+
 coord_fixed()

# correlation between ratios - with rna data
cormat <- round(cor(data_rna_ribo_silac[, c(2:9)], use = "complete.obs"),2)
upper_tri <- get_upper_tri(cormat)

library(reshape2)
melted_cormat <- melt(upper_tri, na.rm = TRUE)
head(melted_cormat)

# Heatmap
library(ggplot2)
plot2 <- ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name="Pearson\nCorrelation") +
  theme_minimal()+ 
        ggtitle("Correlation between foldchanges") +
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1))+
 coord_fixed()

ggarrange(plot1, plot2, labels = c("A", "B"),
          ncol = 2, nrow = 1)

```


### iii) Scatter plots
```{r  message=FALSE, warning=FALSE, fig.width=12, fig.height = 3, exploration1}
data <- rna_ribo_mapped

plot1 <- ggplot(data, aes(x=log2(ribo_0h), y=log2(ribo_4h))) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  labs(ylab("RiboSeq 0h")) +
  labs(xlab("RiboSeq 4h"))
  
plot2 <- ggplot(data, aes(x=log2(ribo_0h), y=log2(ribo_8h))) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  labs(ylab("RiboSeq 0h")) +
  labs(xlab("RiboSeq 8h"))

plot3 <- ggplot(data, aes(x=log2(ribo_0h), y=log2(ribo_14h))) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  labs(ylab("RiboSeq 0h")) +
  labs(xlab("RiboSeq 14h")) 

ggarrange(plot1, plot2, plot3, labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)

cor(data_sub$rna.ratio.4h, data_sub$ribo.ratio.4h, use = "complete.obs", method = "pearson")
# 0.3651829
cor(data_sub$rna.ratio.8h, data_sub$ribo.ratio.8h, use = "complete.obs", method = "pearson")
# 0.5680526
cor(data_sub$rna.ratio.14h, data_sub$ribo.ratio.14h, use = "complete.obs", method = "pearson")
# 0.7960264
```

```{r  message=FALSE, warning=FALSE, fig.width=12, fig.height = 4, exploration2}

plot1 <- ggplot(data_sub, aes(x=rna.ratio.4h, y=ribo.ratio.4h)) + 
  geom_point(color = "#10101020") +
#  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 4 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 4 hours")) 

plot2 <- ggplot(data_sub, aes(x=rna.ratio.8h, y=ribo.ratio.8h)) + 
  geom_point(color = "#10101020") +
#  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 8 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 8 hours")) 

plot3 <- ggplot(data_sub, aes(x=rna.ratio.14h, y=ribo.ratio.14h)) + 
  geom_point(color = "#10101020") +
#  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 14 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 14 hours")) 

ggarrange(plot1, plot2, plot3, labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)
```

```{r  message=FALSE, warning=FALSE, fig.width=12, fig.height = 3, exploration3}

plot1 <- ggplot(data_sub, aes(x=rna.ratio.4h, y=ribo.ratio.4h)) + 
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 4 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 4 hours")) 

plot2 <- ggplot(data_sub, aes(x=rna.ratio.8h, y=ribo.ratio.8h)) + 
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 8 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 8 hours")) 

plot3 <- ggplot(data_sub, aes(x=rna.ratio.14h, y=ribo.ratio.14h)) + 
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 14 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 14 hours")) 

ggarrange(plot1, plot2, plot3, labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)

# scatter plots comparing rna, ribo and silac
plot1 <- ggplot(data_rna_ribo_silac, aes(x=iDUX4_logFC, y=rna.ratio.14h)) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 14 hours - Suja")) +
  labs(xlab("log2 fold change of RNA at 14 hours - Dyle")) 

plot2 <- ggplot(data_rna_ribo_silac, aes(x=rna.ratio.14h, y=ribo.ratio.14h)) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of RNA at 14 hours")) +
  labs(xlab("log2 fold change of Ribosome footprints at 14 hours")) 

plot3 <- ggplot(data_rna_ribo_silac, aes(x=ribo.ratio.14h, y=hl.ratio)) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of Ribosome footprints at 14 hours")) +
  labs(xlab("log2 fold change of protein abundance at 14 hours (SILAC)")) 

ggarrange(plot1, plot2, plot3, labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)

# scatter plots comparing riboSeq time points to SILAC
plot1 <- ggplot(data_rna_ribo_silac, aes(x=ribo.ratio.4h, y=hl.ratio)) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of Ribosome footprints at 4 hours")) +
  labs(xlab("log2 fold change of protein abundance at 14 hours (SILAC)")) 

plot2 <- ggplot(data_rna_ribo_silac, aes(x=ribo.ratio.8h, y=hl.ratio)) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of Ribosome footprints at 8 hours")) +
  labs(xlab("log2 fold change of protein abundance at 14 hours (SILAC)"))  

plot3 <- ggplot(data_rna_ribo_silac, aes(x=ribo.ratio.14h, y=hl.ratio)) + 
#  geom_point(color = "#10101040") +
  geom_hex(bins = 50) +
  theme_few() +
  xlim(-15, 15) +
  ylim(-15, 15) +
  geom_vline(xintercept = 0, color = "#10101050") +
  geom_hline(yintercept = 0, color = "#10101050") +
  geom_abline(intercept = 0, slope = 1, color = "#10101050") +
  labs(ylab("log2 fold change of Ribosome footprints at 14 hours")) +
  labs(xlab("log2 fold change of protein abundance at 14 hours (SILAC)"))  

ggarrange(plot1, plot2, plot3, labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)

```

## More analyses to do

1. Normalize the read coverage for IGV sessions
2. Transcriptional regulation. i.e. RNA, ribo, and protein, all agree. 
3. Translational regulation. i.e. RNA and protein discordant. Protein and ribo concordant. So, protein change is a result of change in translation. 
4. Post-translational regulation. i.e. RNA and ribo changes are concordant. But protein is discordant.
5. Isoform switching - need to do splicing analysis for this. 

